packages/astro/src/container/pipeline.ts

```diff
@@ -0,0 +1,143 @@
+			'development',
+			renderers,
+			resolve,
+			serverLike,
+			streaming
+		);
+	}
+
+	componentMetadata(_routeData: RouteData): Promise<SSRResult['componentMetadata']> | void {}
+
+	headElements(routeData: RouteData): Promise<HeadElements> | HeadElements {
+		const routeInfo = this.manifest.routes.find((route) => route.routeData === routeData);
+		const links = new Set<never>();
+		const scripts = new Set<SSRElement>();
+		const styles = createStylesheetElementSet(routeInfo?.styles ?? []);
+
+		for (const script of routeInfo?.scripts ?? []) {
+			if ('stage' in script) {
+				if (script.stage === 'head-inline') {
+					scripts.add({
+						props: {},
+						children: script.children,
+					});
+				}
+			} else {
+				scripts.add(createModuleScriptElement(script));
+			}
+		}
+		return { links, styles, scripts };
+	}
+
+	async tryRewrite(rewritePayload: RewritePayload): Promise<[RouteData, ComponentInstance]> {
+		let foundRoute: RouteData | undefined;
+		// options.manifest is the actual type that contains the information
+		for (const route of this.manifest.routes) {
+			const routeData = route.routeData;
+			if (rewritePayload instanceof URL) {
+				if (routeData.pattern.test(rewritePayload.pathname)) {
+					foundRoute = routeData;
+					break;
+				}
+			} else if (rewritePayload instanceof Request) {
+				const url = new URL(rewritePayload.url);
+				if (routeData.pattern.test(url.pathname)) {
+					foundRoute = routeData;
+					break;
+				}
+			} else if (routeData.pattern.test(decodeURI(rewritePayload))) {
+				foundRoute = routeData;
+				break;
+			}
+		}
+		if (foundRoute) {
+			const componentInstance = await this.getComponentByRoute(foundRoute);
+			return [foundRoute, componentInstance];
+		} else {
+			throw new AstroError(RouteNotFound);
+		}
+	}
+
+	insertRoute(route: RouteData, componentInstance: ComponentInstance): void {
+		this.#componentsInterner.set(route, {
+			page() {
+				return Promise.resolve(componentInstance);
+			},
+			renderers: this.manifest.renderers,
+			onRequest: this.manifest.middleware,
+		});
+	}
+
+	async getComponentByRoute(routeData: RouteData): Promise<ComponentInstance> {
+		if (this.#componentsInterner.has(routeData)) {
+			// SAFETY: checked before
+			const entry = this.#componentsInterner.get(routeData)!;
+			return await entry.page();
+		} else {
+			// SAFETY: the pipeline calls `retrieveRoutesToGenerate`, which is in charge to fill the cache.
+			const filePath = this.#routesByFilePath.get(routeData)!;
+			const module = await this.#retrieveSsrEntry(routeData, filePath);
+			return module.page();
+		}
+	}
+
+	async #retrieveSsrEntry(route: RouteData, filePath: string): Promise<SinglePageBuiltModule> {
+		if (this.#componentsInterner.has(route)) {
+			// SAFETY: it is checked inside the if
+			return this.#componentsInterner.get(route)!;
+		}
+		const ssrEntryURLPage = this.#createEntryURL(filePath, new URL(process.cwd()));
+		const entry = await import(ssrEntryURLPage.toString());
```

Comment:
this scares me a bit and probably won't work when this is being used in Vite.

Instead of us importing components what do you think about adding an API to allow you to register components inside of a container. That way you can register the ones that might be rewritten to.